indra/llxml/llxmlnode.cpp

```diff
@@ -650,29 +650,52 @@ bool LLXMLNode::updateNode(
 	return true;
 }
 
+static std::map<std::string, LLXMLNodePtr> sXMLCache;
+static LLSharedMutex sXMLCacheMutex;
+
 // static
 bool LLXMLNode::parseFile(const std::string& filename, LLXMLNodePtr& node, LLXMLNode* defaults_tree)
 {
-	LL_DEBUGS("XMLNode") << "parsing XML file: " << filename << LL_ENDL;
+    // XML files root folder: e.g., "C:\\Program Files\\SecondLifeViewer"
+    std::string app_dir = gDirUtilp->getAppRODataDir();
+
+    // True if filename starts with "C:\\Program Files\\SecondLifeViewer"
+    bool embedded = filename.substr(0, app_dir.size()) == app_dir;
+    std::string file_id;
+    if (embedded)
+    {
+        // The rest of filename is XML file unique id
+        // E.g., "app_settings\\settings_files.xml"
+        file_id = filename.substr(app_dir.size() + 1);
+
+        LLSharedMutexLock lock(&sXMLCacheMutex);
+        auto it = sXMLCache.find(file_id);
+        if (it != sXMLCache.end())
+        {
+            node = it->second.notNull() ? it->second->deepCopy() : nullptr;
+            return node.notNull();
+        }
+    }
+
+    LL_DEBUGS("XMLNode") << "parsing XML file: " << filename << LL_ENDL;
```

Comment:
I have a feeling this is going to break XUI preview tool since it relies on re-reading the same file.

Reply:
Fixed